File: funcscript/parser/syntax/FuncScriptParser.GetOperator.cs
using FuncScript.Block;
using FuncScript.Funcs.Math;
using FuncScript.Model;

namespace FuncScript.Core
{
    public partial class FuncScriptParser
    {
        record GetOperatorResult(string MatchedOp, ParseNode ParseNode, int NextIndex)
            : ParseResult(ParseNode, NextIndex);

        static GetOperatorResult GetOperator(ParseContext context, string[] candidates, int index)
        {
            foreach (var op in candidates)
            {
                var literalMatchResult = GetLiteralMatch(context, index, op);
                var i = literalMatchResult.NextIndex;
                if (i <= index) continue;

                var parseNode = new ParseNode(ParseNodeType.Operator, index, i - index);
                return new GetOperatorResult(op, parseNode, i);
            }

            return new GetOperatorResult(null, null, index);
        }
    }
}

File: ts/core/src/parser/syntax/FuncScriptParser.GetOperator.ts
import { ParseContext, ParseNode, ParseNodeType, ParseResult } from "../FuncScriptParser.Main";
import { GetLiteralMatch } from "./FuncScriptParser.GetLiteralMatch";

interface GetOperatorResult extends ParseResult {
    MatchedOp: string | null;
}

export function GetOperator(context: ParseContext, candidates: string[], index: number): GetOperatorResult {
    for (const op of candidates) {
        const literalMatchResult = GetLiteralMatch(context, index, op);
        const i = literalMatchResult.NextIndex;
        if (i <= index) continue;

        const parseNode = new ParseNode(ParseNodeType.Operator, index, i - index);
        return {
            MatchedOp: op,
            ParseNode: parseNode,
            NextIndex: i
        };
    }

    return {
        MatchedOp: null,
        ParseNode: null,
        NextIndex: index
    };
}